aggregator_api/src/lib.rs

```diff
@@ -0,0 +1,897 @@
+                            },
+                        )
+                        .await
+                    }
+                    .instrument(info_span!(ENDPOINT_NAME))
+                }
+            })
+            .delete("/tasks/:task_id", {
+                const ENDPOINT_NAME: &str = "delete_task";
+                |conn: Conn| {
+                    async move {
+                        endpoint_handler::<C, _>(
+                            conn,
+                            ENDPOINT_NAME,
+                            |aggregator_api, conn| async move {
+                                aggregator_api.handle_delete_task(conn).await
+                            },
+                        )
+                        .await
+                    }
+                    .instrument(info_span!(ENDPOINT_NAME))
+                }
+            })
+            .get("/tasks/:task_id/metrics", {
+                const ENDPOINT_NAME: &str = "get_task_metrics";
+                |conn: Conn| {
+                    async move {
+                        endpoint_handler::<C, _>(
+                            conn,
+                            ENDPOINT_NAME,
+                            |aggregator_api, conn| async move {
+                                aggregator_api.handle_get_task_metrics(conn).await
+                            },
+                        )
+                        .await
+                    }
+                    .instrument(info_span!(ENDPOINT_NAME))
+                }
+            }),
+    )
+}
+
+async fn endpoint_handler<C: Clock, Fut>(
+    mut conn: Conn,
+    endpoint_name: impl Into<String>,
+    handler: impl Fn(Arc<AggregatorApi<C>>, Conn) -> Fut,
+) -> Conn
+where
+    Fut: Future<Output = Conn>,
+{
+    let start_instant = Instant::now();
+    let aggregator_api: Arc<AggregatorApi<C>> =
+        Arc::clone(conn.state().expect("No AggregatorApi in connection state"));
+
+    conn = aggregator_api.check_auth(conn);
+    if !conn.is_halted() {
+        conn = handler(Arc::clone(&aggregator_api), conn).await;
+    }
+
+    let status: u16 = match conn.status() {
+        Some(status) => status.into(),
+        None => {
+            warn!(path = conn.path(), "No status set on connection");
+            return conn;
+        }
+    };
+
+    aggregator_api.response_time_histogram.record(
+        &Context::current(),
+        start_instant.elapsed().as_secs_f64(),
+        &[
+            KeyValue::new("endpoint", endpoint_name.into()),
+            KeyValue::new("status", status.to_string()),
+        ],
+    );
+    conn
+}
+
+struct AggregatorApi<C: Clock> {
+    ds: Datastore<C>,
+    cfg: Config,
+    response_time_histogram: Histogram<f64>,
+}
+
+impl<C: Clock> AggregatorApi<C> {
+    fn check_auth(&self, conn: Conn) -> Conn {
+        if let Some(authorization_value) = conn.headers().get("authorization") {
+            if let Some(received_token) = authorization_value.as_ref().strip_prefix(b"Basic ") {
```

Comment:
It doesn't really matter because we control both ends and don't expect this API to generally be interoperable, but I think `Authorization: Bearer <token>` is what we want. `Authorization: Basic` is meant to be used with username/password schemes ([RFC 7617](https://www.rfc-editor.org/rfc/rfc7617)).